zerver/actions/streams.py

```diff
@@ -110,6 +114,13 @@ def do_deactivate_stream(stream: Stream, *, acting_user: Optional[UserProfile])
     stream.name = new_name[: Stream.MAX_NAME_LENGTH]
     stream.save(update_fields=["name", "deactivated", "invite_only"])
 
+    assert stream.recipient_id is not None
+    # Unset the is_web_public cache on attachments, since the stream is now private.
+    Attachment.objects.filter(messages__recipient_id=stream.recipient_id).update(is_web_public=None)
+    ArchivedAttachment.objects.filter(messages__recipient_id=stream.recipient_id).update(
+        is_web_public=None
+    )
```

Comment:
I am not sure whether we want to unset `is_realm_public` field as well. Added a commit to do so when unarchiving a stream though as we make the stream private while unarchiving.

Reply:
I think it would be correct do so. But we probably do want a check for whether the stream was previously public or web-public around these blocks in `do_deactivate_stream` -- since it the stream was already private when deactivated, it should not need to perturb these caches.